whisper.cpp

```diff
@@ -2424,40 +2422,45 @@ static void fft(const std::vector<float> & in, std::vector<float> & out) {
 
         out[2*(k + N/2) + 0] = even_fft[2*k + 0] - re*re_odd + im*im_odd;
         out[2*(k + N/2) + 1] = even_fft[2*k + 1] - re*im_odd - im*re_odd;
+
     }
 }
 
 static void log_mel_spectrogram_worker_thread(int ith, const std::vector<float> &hann, const float *samples,
                                               int n_samples, int fft_size, int fft_step, int n_threads,
                                               const whisper_filters &filters, bool speed_up, whisper_mel &mel) {
-    std::vector<float> fft_in(fft_size, 0.0);
-    std::vector<float> fft_out(2 * fft_size);
     int n_fft = 1 + (speed_up ? fft_size / 4 : fft_size / 2);
 
     for (int i = ith; i < mel.n_len; i += n_threads) {
+        std::vector<float> fft_in(fft_size, 0.0);
+        std::vector<float> fft_out(2 * fft_size);
```

Comment:
What is the benefit of moving these vectors inside the loop?

Reply:
After implementing performance testing with `ggml_time_us();`, I found that the current code segment runs approximately `10% faster` than before. The average running time has improved from `10154μs` to `9127μs`.
https://github.com/ggerganov/whisper.cpp/blob/f3e7774ceea0caa776e785b9c50bbbf1a33515dd/whisper.cpp#L2439-L2447

I do have a question, though. I'm unsure as to how `(n_samples - offset)` can be negative. If that were the case, `fft_in` would simply consist of zeroes. And if `fft_in` is entirely zeroes, why would we need to compute the FFT at all?

> 

<img width="112" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/129547291/0bc14b6b-303b-4666-bffc-09acee21e39a">